src/Catalyst.Core.Modules.Dfs/CoreApi/PinApi.cs

```diff
@@ -0,0 +1,178 @@
+                }
+                
+                var folder = Path.Combine(_repositoryOptions.Folder, "pins");
+                if (!Directory.Exists(folder))
+                {
+                    Directory.CreateDirectory(folder);
+                }
+                
+                _store = new FileStore<Cid, Pin>
+                {
+                    Folder = folder,
+                    NameToKey = cid => cid.Hash.ToBase32(),
+                    KeyToName = key => new MultiHash(key.FromBase32())
+                };
+
+                return _store;
+            }
+        }
+
+        public async Task<IEnumerable<Cid>> AddAsync(string path,
+            bool recursive = true,
+            CancellationToken cancel = default)
+        {
+            var r = await _nameApi.ResolveAsync(path, cancel: cancel).ConfigureAwait(false);
+            var id = Cid.Decode(r.Remove(0, 6));
+            var todos = new Stack<Cid>();
+            todos.Push(id);
+            var dones = new List<Cid>();
+
+            // The pin is added before the content is fetched, so that
+            // garbage collection will not delete the newly pinned
+            // content.
+
+            while (todos.Count > 0)
+            {
+                var current = todos.Pop();
+
+                // Add CID to PIN database.
+                await Store.PutAsync(current, new Pin
+                {
+                    Id = current
+                }, cancel).ConfigureAwait(false);
+
+                // Make sure that the content is stored locally.
+                await BlockApi.GetAsync(current, cancel).ConfigureAwait(false);
+
+                // Recursively pin the links?
+                if (recursive && current.ContentType == "dag-pb")
+                {
+                    var links = await _objectApi.LinksAsync(current, cancel);
+                    foreach (var link in links)
+                    {
+                        todos.Push(link.Id);
+                    }
+                }
+
+                dones.Add(current);
+            }
+
+            return dones;
+        }
+
+        public Task<IEnumerable<Cid>> ListAsync(CancellationToken cancel = default)
+        {
+            var cids = Store.Values.Select(pin => pin.Id);
+            return Task.FromResult(cids);
+        }
+
+        public async Task<IEnumerable<Cid>> RemoveAsync(Cid id,
+            bool recursive = true,
+            CancellationToken cancel = default)
+        {
+            var todos = new Stack<Cid>();
+            todos.Push(id);
+            var dones = new List<Cid>();
+
+            while (todos.Count > 0)
+            {
+                var current = todos.Pop();
+                await Store.RemoveAsync(current, cancel).ConfigureAwait(false);
+                if (recursive)
+                {
+                    if (null != await BlockApi.StatAsync(current, cancel).ConfigureAwait(false))
+                    {
+                        try
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this code to not nest more than 3 control flow statements.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)